capsules/src/adc.rs

```diff
@@ -1,77 +1,777 @@
 //! ADC Capsule
 //!
 //! Provides userspace applications with the ability to sample
-//! ADC channels.
+//! analog signals.
 
 use core::cell::Cell;
-use kernel::{AppId, Callback, Container, Driver, ReturnCode};
-use kernel::hil::adc::{Client, AdcSingle};
+use core::cmp;
+use kernel::{AppId, AppSlice, Callback, Driver, ReturnCode, Shared};
+use kernel::common::take_cell::{MapCell, TakeCell};
+use kernel::hil;
 
-#[derive(Default)]
-pub struct AppData {
-    channel: Option<u8>,
-    callback: Option<Callback>,
+/// ADC application driver, used by applications to interact with ADC.
+/// Not currently virtualized, only one application can use it at a time.
+pub struct Adc<'a, A: hil::adc::Adc + hil::adc::AdcHighSpeed + 'a> {
+    // ADC driver
+    adc: &'a A,
+    channels: &'a [&'a <A as hil::adc::Adc>::Channel],
+
+    // ADC state
+    active: Cell<bool>,
+    mode: Cell<AdcMode>,
+
+    // App state
+    app_state: MapCell<AppState>,
+    channel: Cell<usize>,
+    callback: Cell<Option<Callback>>,
+    app_buf_offset: Cell<usize>,
+    samples_remaining: Cell<usize>,
+    samples_outstanding: Cell<usize>,
+    next_samples_outstanding: Cell<usize>,
+    using_app_buf1: Cell<bool>,
+
+    // ADC buffers
+    adc_buf1: TakeCell<'static, [u16]>,
+    adc_buf2: TakeCell<'static, [u16]>,
+    adc_buf3: TakeCell<'static, [u16]>,
 }
 
-pub struct ADC<'a, A: AdcSingle + 'a> {
-    adc: &'a A,
-    channel: Cell<Option<u8>>,
-    app: Container<AppData>,
+/// ADC modes, used to track internal state and to signify to applications which
+/// state a callback came from
+#[derive(Copy,Clone,PartialEq)]
```

Comment:
/ducks

... since others have asked for some changes, toss a `Debug` in this derive list?